### PR TITLE
added usage tracking for symbols

### LINKED ISSUE #104: Show warning when a symbol is not used

### PR DESCRIPTION
**Pull request by gerdreiss**
closes #104

Hi @hrj , please let me know whether I'm on the right path with this solution. 
There is also a possibility to keep a usage state within Definition, but I've chosen this way as I find it less intrusive, and the tracking is kept central this way.

### CODE REVIEW

**Comment by hrj** [comment]
Quick questions (I am travelling):
* Have you tried using a set for the `usedDefinitions` instead of List? `Set` seems more appropriate
* I think the test should check for stdout / stderr so that the warning is actually covered. Not sure whether the test framework supports that. CC @jaa127  
* Have you verified that full path of the definition is being checked? You can add that scenario in the test.

**Comment by gerdreiss** [comment]
@hrj Agreed to using Set instead of List - it is more appropriate, indeed. The new commit contains the corresponding change. I'll look into the full path check next.

**Comment by gerdreiss** [comment]
@hrj I've reworked my solution so that definitions themselves have a state now which is set to "used" whenever it's used in an evaluation. The check for the unused defs is now in the class Scope where I just filter all definitions defined in the given scope including the nested scopes for those that are unused, and print a warning if any found.

File: base/src/main/scala/co/uproot/abandon/Ast.scala [hunk 1]
```diff
@@ -272,8 +272,10 @@ sealed class ASTTangibleEntry extends ASTEntry
 
 case class Transaction(pos: InputPosition, date: Date, posts: Seq[Post], annotationOpt: Option[String], payeeOpt: Option[String], comments: List[String]) extends ASTTangibleEntry
 
-case class Definition(pos: InputPosition, name: String, params: List[String], rhs: Expr) extends ASTTangibleEntry {
-  def prettyPrint = "def %s(%s) = %s" format (name, params.mkString(", "), rhs.prettyPrint)
+case class Definition(pos: InputPosition, name: String, params: List[String], rhs: Expr, private var used: Boolean = false) extends ASTTangibleEntry {
```
Comment (hrj) [inline base/src/main/scala/co/uproot/abandon/Ast.scala]: Could the `used` constructor parameter be changed to a member variable? If it's part of the constructor parameter list, it becomes a part of its identity.

File: base/src/main/scala/co/uproot/abandon/Ast.scala [hunk 2]
```diff
@@ -272,8 +272,10 @@ sealed class ASTTangibleEntry extends ASTEntry
...
+  def markAsUsed(): Unit = used = true
+  def isUsed: Boolean = used
+  def prettyPrint(): Unit = "def %s(%s) = %s" format (name, params.mkString(", "), rhs.prettyPrint)
```
Comment (hrj) [inline base/src/main/scala/co/uproot/abandon/Ast.scala]: You changed the return type of `prettyPrint` to `Unit`. Was that intentional?

File: base/src/main/scala/co/uproot/abandon/Ast.scala [hunk 3]
```diff
@@ -369,4 +371,9 @@ case class Scope(entries: Seq[ASTEntry], parentOpt: Option[Scope]) extends ASTEn
         includedScopes.foreach(_.checkDupes())
     }
   }
+
+  def checkUnusedSymbols() {
+    definitions.filterNot(_.isUsed)
+      .foreach(d => println(s"symbol `${d.name}` defined in ${d.pos} but never used"))
```
Comment (hrj) [inline base/src/main/scala/co/uproot/abandon/Ast.scala]: This should perhaps be printed to `stderr`, not `stdout`. But I am not sure if the rest of the code shows warnings / errors to `stderr`. Ignore this comment if not.

File: tests/sclT0009-unused-symbols/sym01.exec [hunk 4]
```diff
@@ -0,0 +1,2 @@
+# format: exec
+exec:
```
Comment (hrj) [inline tests/sclT0009-unused-symbols/sym01.exec]: This looks like a great test :+1: 
But it doesn't check whether the output matches the expected value?

File: base/src/main/scala/co/uproot/abandon/Ast.scala (hunk 2 above)
Comment (gerdreiss) [inline base/src/main/scala/co/uproot/abandon/Ast.scala]: That was intentional. And stupid. That's why code reviews are so important ;-)

File: base/src/main/scala/co/uproot/abandon/Ast.scala (hunk 1 above)
Comment (gerdreiss) [inline base/src/main/scala/co/uproot/abandon/Ast.scala]: You're right, I'll change that.

File: base/src/main/scala/co/uproot/abandon/Ast.scala (hunk 3 above)
Comment (gerdreiss) [inline base/src/main/scala/co/uproot/abandon/Ast.scala]: As far as I've seen all errors are written in stdout using println, except one occasion in UI. But there is one print out using formatting for the printed string:

`println(Console.RED + Console.BOLD + msg + Console.RESET)`

I think that's a good way to print a warning, but maybe using another colour?

File: tests/sclT0009-unused-symbols/sym01.exec (hunk 4 above)
Comment (gerdreiss) [inline tests/sclT0009-unused-symbols/sym01.exec]: I'll add that part.

Comment (hrj) [inline tests/sclT0009-unused-symbols/sym01.exec]: Oh, what I meant was: the unused symbol warning isn't being checked, AFAICT.

Comment (gerdreiss) [inline tests/sclT0009-unused-symbols/sym01.exec]: Well, tbh, I wouldn't know how to test the println within these tests. What I could do though is to test the warnings in the ProcessorTest using the old Java trick:

`final ByteArrayOutputStream myOut = new ByteArrayOutputStream();
System.setOut(new PrintStream(myOut));

// test stuff here...

final String standardOutput = myOut.toString();

//test standardOutput here
`

What do you think?

Comment (hrj) [inline tests/sclT0009-unused-symbols/sym01.exec]: I have asked a question here: https://github.com/sn127/utils/issues/16

Otherwise, what you describe sounds good.

File: base/src/main/scala/co/uproot/abandon/Ast.scala (hunk 3 above)
Comment (hrj) [inline base/src/main/scala/co/uproot/abandon/Ast.scala]: > I think that's a good way to print a warning, but maybe using another colour?

Yeah, that sounds good. Orange or Yellow would be nice.

File: tests/sclT0009-unused-symbols/sym01.exec (hunk 4 above)
Comment (hrj) [inline tests/sclT0009-unused-symbols/sym01.exec]: Oh wow! I was going to suggest adding the stdout test later. But you have done it already! :clap: 

Will do a final review tomorrow before merging.

Comment (gerdreiss) [inline tests/sclT0009-unused-symbols/sym01.exec]: Yes, but the tests are failing now on Travis, although they are running fine locally :(
I'm not sure what the problem is...

Comment (jaa127) [inline tests/sclT0009-unused-symbols/sym01.exec]: Congrats - this is cool!

Tests are falling because there is abs path of your home + working directory in reference file.
Now for that problem, there are few options:
 1) Modify Abandon, so that it won't print abs path (not so good)
 2) Modify DirSuite a little bit more: https://github.com/sn127/utils/blob/master/docs/dirsuite.md#selecting-validator

So you can overload / provide specialized selectValidator which can do fuzzy validation, and ignore changing path.  These are really simple functions, hard part will be to implement fuzzy-validator:

https://github.com/sn127/utils/blob/master/testing/src/main/scala/fi/sn127/utils/testing/DirSuiteLike.scala#L284

https://github.com/sn127/utils/blob/master/testing/src/main/scala/fi/sn127/utils/testing/TestValidator.scala

Comment (hrj) [inline tests/sclT0009-unused-symbols/sym01.exec]: @jaa127 Thanks for your help!

I think this test is getting too complex, so we could move it to a separate PR. For this PR I would be happy if the warning is emitted and checked via unit test. The system test using CLI and stdout need not be part of this PR.

Comment (gerdreiss) [inline tests/sclT0009-unused-symbols/sym01.exec]: > Tests are falling because there is abs path of your home + working directory in reference file.

Oh god, this is embarrassing :/ 

@jaa127 I'll refine my solution following your advice, but in the new PR, as @hrj suggests.

@hrj I'm going to take the tests out of this PR, and create a new one for the tests.

**Comment by hrj** [comment]
Was superseded by #147
